// typed.ts/lib/esm/types/queue.js
export class Queue {
    constructor(resetter) {
        this._items = [];
        this._queueIndex = 0;
        this._queueDetailIndex = 0;
        this._resetter = resetter;
    }
    add(item) {
        this._items.push(item);
    }
    clear() {
        this._items.length = 0;
        this.resetIndices();
    }
    resetIndices() {
        this._queueIndex = 0;
        this._queueDetailIndex = 0;
    }
    get item() {
        return this._items[this._queueIndex];
    }
    get detailIndex() {
        return this._queueDetailIndex;
    }
    increment(maxDetailIndex) {
        if (this._resetter.isReset) {
            return false;
        }
        this._queueDetailIndex++;
        if (!maxDetailIndex || this._queueDetailIndex === maxDetailIndex) {
            return this.nextQueueItem();
        }
        else {
            return true;
        }
    }
    nextQueueItem() {
        if (this._resetter.isReset) {
            return false;
        }
        this._queueIndex++;
        if (this._queueIndex === this._items.length) {
            return false;
        }
        this._queueDetailIndex = 0;
        return true;
    }
}
